File: range-sum-query-immutable/range-sum-query-immutable.cpp
class NumArray {
public:
    
    int n = 0;
    vector<int> sum;
    // vector<int>& loc_nums;
    
    NumArray(vector<int>& nums){
        // this->loc_nums = nums;
        this->n = nums.size();
        this->sum.resize(n+1);
        
        
        for(int i = 0; i < this->n; ++i) {
            sum[i+1] = sum[i] + nums[i];
        }
        
    }
    
    int sumRange(int left, int right) {
        return sum[right+1] - sum[left];
        
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * int param_1 = obj->sumRange(left,right);
 */

File: range-sum-query-immutable/range-sum-query-immutable.py
class NumArray:

    def __init__(self, nums: List[int]):
        self.nums = nums
        self.dp = []
        self.dp.append(nums[0])
        for i in range(1, len(nums)):
            self.dp.append(nums[i] + self.dp[-1])
            
        

    def sumRange(self, left: int, right: int) -> int:
        return self.dp[right] - self.dp[left] + self.nums[left]
        


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)